### PR TITLE
Implementing StorageAgent (NEW)

### LINKED ISSUE #309: Implementation of Storage agent
Include trigger revoking mechanism

Edit SP:
This probably makes sense to develop on top of #308

### PR DESCRIPTION
**Pull request by sebastian-peter**
Resolves #309

### CODE REVIEW

File: src/main/scala/edu/ie3/simona/model/participant/StorageModel.scala [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+      qControl,
+      sRated,
+      cosPhiRated,
+    ) {
+
+  private val minEnergy = zeroKWH
+
+  /** Tolerance for power comparisons. Amounts to 1 W for 1 GWh storage
+    */
+  private implicit val powerTolerance: Power = eStorage / Seconds(1) / 3.6e12
```
Comment (sebastian-peter) [inline src/main/scala/edu/ie3/simona/model/participant/StorageModel.scala:55]: tl;dr I found out why we want it this way when I typed this out

I have some food for thought here myself. Adapting the powerTolerance to be relative to the capacity has been a todo for some time and has now been solved here. For reference, this value is currently only used when checking whether the charging power is close to zero, which is only checked once here (`setPower ~= zeroKW`).

So now I have some questions. Does it really make sense to base the power tolerance on capacity, and the `toleranceMargin` related to SOC based on the maximum power? I feel like both should be rather calculated from `pMax`, because I think it's not impossible to have a battery with large storage and rather small charging power and then we should base comparisons of power values on the maximum power, not capacity.

I think the problem that we're trying to solve here is that in a situation where we charge with a very small power, let's say Mikrowatts, the calculation of the tick at which FlexOptions change will result in a very large value. The value might be so large that the `Long` value overflows, which we want to prevent. 

So there's only two things left to do from my side:
- check if the 1 W tolerance for a GWh installation suffices
- enhance the documentation of the value with a short version of the explanation (var overflow)

**Review by jo-bao — CHANGES_REQUESTED** [review]
In a nutshell, it looks good to me! Just a few small comments. Since it is an important PR, someone else should review it.

File: src/main/scala/edu/ie3/simona/agent/participant/storage/StorageAgentFundamentals.scala [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        StorageRelevantData,
+        StorageState,
+        StorageModel,
+      ],
+      StorageState,
+      Dimensionless,
+  ) => ApparentPower =
+    (_, _, _, _) =>
+      throw new InvalidRequestException(
+        "PV model cannot be run without secondary data."
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/agent/participant/storage/StorageAgentFundamentals.scala:198]: Good catch

File: src/main/scala/edu/ie3/simona/agent/participant/storage/StorageAgentFundamentals.scala [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+      val storageResult = new StorageResult(
+        dateTime,
+        uuid,
+        result.p.toMegawatts.asMegaWatt,
+        result.q.toMegavars.asMegaVar,
+        soc,
+      )
+
+      notifyListener(ParticipantResultEvent(storageResult))
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/agent/participant/storage/StorageAgentFundamentals.scala:297]: @jo-bao Why would it?

File: src/test/scala/edu/ie3/simona/test/common/input/StorageInputTestData.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    "Dummy_StorageTypeInput",
+    Quantities.getQuantity(100d, EURO),
+    Quantities.getQuantity(101d, EURO_PER_MEGAWATTHOUR),
+    Quantities.getQuantity(200d, KILOWATTHOUR),
+    Quantities.getQuantity(13d, KILOVOLTAMPERE),
+    0.997,
+    Quantities.getQuantity(12.961, KILOWATT),
+    Quantities.getQuantity(0.03, PU_PER_HOUR),
+    Quantities.getQuantity(0.92, PU),
+  )
```
Comment (danielfeismann) [inline src/test/scala/edu/ie3/simona/test/common/input/StorageInputTestData.scala:33]:
```suggestion
  protected val storageTypeInput = new StorageTypeInput(
    UUID.fromString("fbee4995-24dd-45e4-9c85-7d986fe99ff3"),
    "Dummy_StorageTypeInput",
    Quantities.getQuantity(15000d, EURO),
    Quantities.getQuantity(0.05d, EURO_PER_MEGAWATTHOUR),
    Quantities.getQuantity(20d, KILOWATTHOUR),
    Quantities.getQuantity(10d, KILOVOLTAMPERE),
    0.997,
    Quantities.getQuantity(9.97, KILOWATT),
    Quantities.getQuantity(0.03, PU_PER_HOUR),
    Quantities.getQuantity(0.92, PU),
  )
```
This is very optional, but I would like to provide realistic test data for two reasons: 

1. users / students get a feeling for realistic values
2. we may find bugs faster since values seems to be unrealistic

File: src/main/scala/edu/ie3/simona/model/participant/StorageModel.scala [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+          }
+          .getOrElse(minEnergy)
+
+        val energyToEmpty = currentStoredEnergy - closestEnergyTarget
+        Some(energyToEmpty / (netPower * (-1)))
+      }
+
+    // calculate the tick from time span
+    val maybeNextTick = maybeTimeSpan.map { timeSpan =>
+      val ticksToEmpty = Math.round(timeSpan.toSeconds)
```
Comment (danielfeismann) [inline src/main/scala/edu/ie3/simona/model/participant/StorageModel.scala:257]: Wouldn't it be ticks to empty or full or target / limit?

File: src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+    "Test_StorageTypeInput",
+    getQuantity(100d, EURO),
+    getQuantity(101d, EURO_PER_MEGAWATTHOUR),
+    getQuantity(100d, KILOWATTHOUR),
+    getQuantity(13d, KILOVOLTAMPERE),
+    0.997,
+    getQuantity(10d, KILOWATT),
+    getQuantity(0.03, PU_PER_HOUR),
+    getQuantity(0.9, PU),
+    )
```
Comment (danielfeismann) [inline src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy:57]: see testdata discussion

Comment (danielfeismann) [reply src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy:57]: By the way: Shouldn't we use the testdata here?

File: src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+    // almost discharged to lowest allowed charge
+    10           | -9        | 3590      || 0    | -10  | 10
+    // charged to mid-level charge
+    41           | 10        | 3600      || 0    | -10  | 10
+    // discharged to mid-level charge
+    60           | -9        | 3600      || 0    | -10  | 10
+    // almost fully charged
+    95.5         | 4.98      | 3600      || 0    | -10  | 10
+    // fully charged
+    95.5         | 5         | 3600      || 0    | -10  | 0
```
Comment (danielfeismann) [inline src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy:123]: This is fully charged because with 5 kW over one hour it will be above 100%, right?

**Review by danielfeismann — CHANGES_REQUESTED** [review]
Just some minor questions and dicussion

File: src/main/scala/edu/ie3/simona/config/ConfigFailFast.scala [hunk 8]
```diff
@@ -659,6 +662,43 @@ case object ConfigFailFast extends LazyLogging {
...
+    *   RuntimeConfig of Storages
+    */
+  private def checkStoragesConfig(
+      storageConfig: SimonaConfig.Simona.Runtime.Participant.Storage
+  ): Unit = {
+    if (
+      storageConfig.defaultConfig.initialSoc < 0.0 || storageConfig.defaultConfig.initialSoc > 1.0
+    )
+      throw new RuntimeException(
+        s"StorageRuntimeConfig: Default initial SOC needs to be between 0.0 and 1.0."
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/config/ConfigFailFast.scala:677]: Hmm but `Soc` is weird in prose text? Maybe we should rather adapt rtd?

[the remark this answers is not in the document]
Comment (danielfeismann) [reply src/main/scala/edu/ie3/simona/config/ConfigFailFast.scala:677]: also fine :)

File: src/main/scala/edu/ie3/simona/config/ConfigFailFast.scala [hunk 9]
```diff
@@ -659,6 +662,43 @@ case object ConfigFailFast extends LazyLogging {
...
+      )
+
+    if (
+      storageConfig.defaultConfig.targetSoc.exists(
+        _ < 0.0
+      ) || storageConfig.defaultConfig.targetSoc.exists(_ > 1.0)
+    )
+      throw new RuntimeException(
+        s"StorageRuntimeConfig: Default target SOC needs to be between 0.0 and 1.0."
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/config/ConfigFailFast.scala:686]: See above

File: src/main/scala/edu/ie3/simona/config/ConfigFailFast.scala [hunk 10]
```diff
@@ -659,6 +662,43 @@ case object ConfigFailFast extends LazyLogging {
...
+
+    storageConfig.individualConfigs.foreach { config =>
+      if (config.initialSoc < 0.0 || config.initialSoc > 1.0)
+        throw new RuntimeException(
+          s"StorageRuntimeConfig: ${config.uuids} initial SOC needs to be between 0.0 and 1.0."
+        )
+
+      if (config.targetSoc.exists(_ < 0.0) || config.targetSoc.exists(_ > 1.0))
+        throw new RuntimeException(
+          s"StorageRuntimeConfig: ${config.uuids} target SOC needs to be between 0.0 and 1.0."
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/config/ConfigFailFast.scala:697]: See above

File: src/main/scala/edu/ie3/simona/config/ConfigFailFast.scala [hunk 11]
```diff
@@ -659,6 +662,43 @@ case object ConfigFailFast extends LazyLogging {
...
+      ) || storageConfig.defaultConfig.targetSoc.exists(_ > 1.0)
+    )
+      throw new RuntimeException(
+        s"StorageRuntimeConfig: Default target SOC needs to be between 0.0 and 1.0."
+      )
+
+    storageConfig.individualConfigs.foreach { config =>
+      if (config.initialSoc < 0.0 || config.initialSoc > 1.0)
+        throw new RuntimeException(
+          s"StorageRuntimeConfig: ${config.uuids} initial SOC needs to be between 0.0 and 1.0."
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/config/ConfigFailFast.scala:692]: See above

File: src/main/scala/edu/ie3/simona/model/participant/StorageModel.scala (hunk 5 above)
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/model/participant/StorageModel.scala:257]: Good catch

File: src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy (hunk 7 above)
Comment (sebastian-peter) [reply src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy:123]: No, we're exactly fully charged because with 90% efficiency, we're net charging with 4.5 kW

File: src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy (hunk 6 above)
Comment (sebastian-peter) [reply src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy:57]: It's currently not possible to use the scala test data here in groovy. The groovy test should be converted to a scalatest anyways at some point, but for now it's more urgent to get the StorageAgent into dev, imho

File: src/test/scala/edu/ie3/simona/test/common/input/StorageInputTestData.scala (hunk 4 above)
Comment (sebastian-peter) [reply src/test/scala/edu/ie3/simona/test/common/input/StorageInputTestData.scala:33]: I agree, we'll just have to adapt a bunch of tests then.

File: src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy (hunk 6 above)
Comment (danielfeismann) [reply src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy:57]: Alright, I agree.

File: src/test/scala/edu/ie3/simona/test/common/input/StorageInputTestData.scala (hunk 4 above)
Comment (danielfeismann) [reply src/test/scala/edu/ie3/simona/test/common/input/StorageInputTestData.scala:33]: True, I'm fine with the 200 kWh Storage. Please just adapt the costs :)

File: src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy (hunk 7 above)
Comment (danielfeismann) [reply src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy:123]: Got it, thanks!

File: src/test/scala/edu/ie3/simona/test/common/input/StorageInputTestData.scala (hunk 4 above)
Comment (sebastian-peter) [reply src/test/scala/edu/ie3/simona/test/common/input/StorageInputTestData.scala]: Done

**Review by danielfeismann — APPROVED** [review]
Great achievement, thanks for your work!
